cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/engine/SQLEngine.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright © 2021 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.etl.api.engine;
+
+import io.cdap.cdap.api.annotation.Beta;
+import io.cdap.cdap.api.data.batch.InputFormatProvider;
+import io.cdap.cdap.api.data.batch.OutputFormatProvider;
+import io.cdap.cdap.etl.api.PipelineConfigurable;
+import io.cdap.cdap.etl.api.SubmitterLifecycle;
+import io.cdap.cdap.etl.api.batch.BatchContext;
+import io.cdap.cdap.etl.api.join.JoinDefinition;
+
+/**
+ * A SQL Engine can be used to pushdown certain dataset operations.
+ *
+ * SQL Engines are implemented as plugins.
+ */
+@Beta
+public interface SQLEngine extends PipelineConfigurable, SubmitterLifecycle<BatchContext> {
+  String PLUGIN_TYPE = "sqlengine";
+
+  /**
+   * Creates an Output Format Provided that can be used to push records into a SQL Engine.
+   * @param tableName The name of the table to use to store these records.
+   * @return an {@link OutputFormatProvider} instance that can be used to write records to the SQL Engine into the
+   * specified table.
+   */
+  OutputFormatProvider push(String tableName);
+
+  /**
+   * Creates an InputFormatProvider that can be used to pull records from the specified table
+   * @param tableName the name of the table to pull records from
+   * @return AN {@link InputFormatProvider} instance that can be used to read records form the specified table in the
+   * SQL Engine.
+   */
+  InputFormatProvider pull(String tableName);
+
+  /**
+   * Check if this table exists in the SQL Engine
+   * @param tableName the table name
+   * @return boolean specifying if this table exists in the remote engine.
+   */
+  boolean exists(String tableName);
```

Comment:
We need some more detail on failure scenarios. Seems like this should allow throwing some sort of exception, maybe an IOException.

One thing we've done for the provisioner APIs, is if a method throws a RetryableException, the caller is responsible for retrying for some amount of time with some backoff strategy. I think it would make sense to have the same contract here, unless you think it would make more sense for the engine itself to retry. (For example, does the BQ client automatically retry itself?)

Reply:
I think that contract makes sense.

Ideally, the engine itself should try to "resolve" issues internally, but throwing either a RetryableException for network issues/transient issues or a new SQLEngineException (which contains details about what happened: Credentials, quota limits, out of disk space, etc) if the error is not recoverable/retryable